### PR TITLE
planner: ban usage of `_tidb_softdelete_time` when `tidb_translate_softdelete_sql` is off | tidb-test=feature/active-active tikv=feature/active-active

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref #64281

### What changed and how does it work?

Add a check in `Preprocess` stage: if there exists `_tidb_softdelete_time` in column name anywhere in a `SELECT`/DML SQL, return an error.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/preprocess.go [hunk 1]
```diff
@@ -449,6 +449,16 @@ func (p *preprocessor) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
 	case *ast.Constraint:
 		// Used in ALTER TABLE or CREATE TABLE
 		p.checkConstraintGrammar(node)
+	case *ast.ColumnName:
+		// When tidb_translate_softdelete_sql is enabled, users should not reference the _tidb_softdelete_time
+		// hidden column directly in SELECT/INSERT/UPDATE/DELETE statements.
```
Comment (lcwangchao) [inline pkg/planner/core/preprocess.go]: Better to add some comments on why other statements may allow to use _tidb_softdelete_time

File: pkg/planner/core/preprocess.go [hunk 2]
```diff
@@ -449,6 +449,16 @@ func (p *preprocessor) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
...
+		if p.sctx.GetSessionVars().SoftDeleteRewrite &&
+			node.Name.L == model.ExtraSoftDeleteTimeName.L &&
+			(p.stmtTp == TypeSelect || p.stmtTp == TypeInsert || p.stmtTp == TypeUpdate || p.stmtTp == TypeDelete) {
```
Comment (lcwangchao) [inline pkg/planner/core/preprocess.go]: load_data, import into, may be other statements should also check it . Should we use a whitelist to allow to specify _tidb_softdelete_time when SoftDeleteRewrite is true?

Comment (time-and-fate) [reply pkg/planner/core/preprocess.go]: Added checks, including non-transactional DMLs.
Since the softdelete semantics are only effective for DMLs and `SELECT` statements, which are a small and explicit portion of all kinds of statements, I tend to use a blacklist.

File: pkg/planner/core/preprocess.go (hunk 1 above)
Comment (time-and-fate) [reply pkg/planner/core/preprocess.go]: Added some comments.

File: pkg/planner/core/preprocess.go [hunk 3]
```diff
@@ -449,6 +453,20 @@ func (p *preprocessor) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
...
+		// When tidb_translate_softdelete_sql is enabled, the softdelete semantics are enabled for softdelete tables,
+		// and _tidb_softdelete_time column have special meaning, so user query should not reference the hidden column.
+		// The special semantics only apply to DML and SELECT statements, for which we implemented corresponding
+		// rewriting logic, so we only check these statements here.
+		// The _tidb_softdelete_time column can only appear in softdelete table, so we can just check the column name
+		// here without checking the TableInfo.
+		if p.sctx.GetSessionVars().SoftDeleteRewrite && node.Name.L == model.ExtraSoftDeleteTimeName.L &&
+			(p.stmtTp == TypeSelect || p.stmtTp == TypeInsert || p.stmtTp == TypeUpdate || p.stmtTp == TypeDelete ||
+				p.stmtTp == TypeLoadData || p.stmtTp == TypeImportInto || p.stmtTp == TypeNonTransactionalDML) {
+			p.err = plannererrors.ErrInternal.GenWithStack(
```
Comment (lcwangchao) [inline pkg/planner/core/preprocess.go:466]: Does it also affect the execute statement?

Comment (time-and-fate) [reply pkg/planner/core/preprocess.go:466]: `Preprocess` would be called separately during `PREPARE`, so we don't need to handle `PREPARE/EXECUTE` here. I added a test case.

**Review by lcwangchao — APPROVED** [review]

File: pkg/planner/core/preprocess.go (hunk 3 above)
Comment (time-and-fate) [reply pkg/planner/core/preprocess.go:466]: Due to #65266, an edge case for `PREPARE/EXECUTE` is not handled correctly. I also added them to the test case.

**Review by winoros — APPROVED** [review]
